presto-main/src/main/java/com/facebook/presto/operator/GroupIdOperator.java

```diff
@@ -0,0 +1,260 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableSet;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class GroupIdOperator
+        implements Operator
+{
+    public static class GroupIdOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final List<Type> types;
+        private final List<List<Integer>> groupingSetChannels;
+
+        private boolean closed;
+
+        public GroupIdOperatorFactory(
+                int operatorId,
+                List<? extends Type> inputTypes,
+                List<List<Integer>> groupingSetChannels)
+        {
+            this.operatorId = operatorId;
+            this.groupingSetChannels = ImmutableList.copyOf(groupingSetChannels);
+
+            // add the groupId channel to the output types
+            this.types = ImmutableList.<Type>builder().add(BIGINT).addAll(inputTypes).build();
+        }
+
+        @Override
+        public List<Type> getTypes()
+        {
+            return types;
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+            OperatorContext operatorContext = driverContext.addOperatorContext(operatorId, GroupIdOperator.class.getSimpleName());
+
+            return new GroupIdOperator(
+                    operatorContext,
+                    types,
+                    groupingSetChannels);
+        }
+
+        @Override
+        public void close()
+        {
+            closed = true;
+        }
+    }
+
+    private final OperatorContext operatorContext;
+
+    private final List<Type> types;
+
+    private GroupIdBuilder groupIdBuilder;
+    private Iterator<Page> outputIterator;
```

Comment:
Use `Collections.emptyIterator()` instead of nulls. This will should simplify `getOutput` quite a bit.